vizro-ai/src/vizro_ai/dashboard/_response_models/page.py

```diff
@@ -0,0 +1,222 @@
+            raise ValidationError(f"Component ids must be unique. Duplicated component ids: {duplicates}")
+        return values
+
+    def __init__(self, **data):
+        """Initialize the page plan."""
+        super().__init__(**data)
+        self._components = None
+        self._controls = None
+        self._layout = None
+
+    def _get_components(self, model, all_df_metadata):
+        if self._components is None:
+            self._components = self._build_components(model=model, all_df_metadata=all_df_metadata)
+        return self._components
+
+    def _build_components(self, model, all_df_metadata):
+        components = []
+        component_log = tqdm(total=0, bar_format="{desc}", leave=False)
+        with tqdm(
+            total=len(self.components_plan),
+            desc=f"Currently Building ... [Page] <{self.title}> components",
+            leave=False,
+        ) as pbar:
+            for component_plan in self.components_plan:
+                component_log.set_description_str(f"[Page] <{self.title}>: [Component] {component_plan.component_id}")
+                pbar.update(1)
+                components.append(component_plan.create(model=model, all_df_metadata=all_df_metadata))
+        component_log.close()
+        return components
+
+    def _get_layout(self, model, all_df_metadata):
+        if self._layout is None:
+            self._layout = self._build_layout(model, all_df_metadata)
+        return self._layout
+
+    def _build_layout(self, model, all_df_metadata):
+        if self.layout_plan is None:
+            return None
+        return self.layout_plan.create(
+            component_ids=self._get_component_ids(model=model, all_df_metadata=all_df_metadata),
+        )
+
+    def _get_controls(self, model, all_df_metadata):
+        if self._controls is None:
+            self._controls = self._build_controls(model=model, all_df_metadata=all_df_metadata)
+        return self._controls
+
+    def _controllable_components(self, model, all_df_metadata):
+        return [
+            comp.id
+            for comp in self._get_components(model=model, all_df_metadata=all_df_metadata)
+            if isinstance(comp, (vm.Graph, vm.AgGrid))
+        ]
+
+    def _get_component_ids(self, model, all_df_metadata):
+        return [comp.id for comp in self._get_components(model=model, all_df_metadata=all_df_metadata)]
+
+    def _build_controls(self, model, all_df_metadata):
+        controls = []
+        with tqdm(
+            total=len(self.controls_plan),
+            desc=f"Currently Building ... [Page] <{self.title}> controls",
+            leave=False,
+        ) as pbar:
+            for control_plan in self.controls_plan:
+                pbar.update(1)
+                control = control_plan.create(
+                    model=model,
+                    controllable_components=self._controllable_components(model=model, all_df_metadata=all_df_metadata),
+                    all_df_metadata=all_df_metadata,
```

Comment:
I am not able to completely digest this at the moment, but we might want to discuss this a little further again. I think it is correct in passing all data, but should we not then infer from the decided targets what columns can be filtered, and offer that to the filter model?

Reply:
Haha, I don't quite understand your question still. Let's discuss further and make updates on next PR if needed. 👍